Document about jsonschema for dvc.yaml (#1650)
iterative/dvc.org: content/docs/user-guide/dvc-files-and-directories.md

`dvc.yaml` files also support `# comments`.

We have published the

> [jsonschema for `dvc.yaml`](https://github.com/iterative/pipeline-schema) files

Reviewer: ```suggestion
[JSON Schema](https://github.com/iterative/pipeline-schema) for `dvc.yaml` files
```

that helps code editors and IDEs to provide auto-completion and

> validation/linting and to verify proper structure of `dvc.yaml` locally and/or

Reviewer: ```suggestion
validation/linting, and to verify proper structure of `dvc.yaml` locally and/or
```

> [schema](https://github.com/iterative/dvcyaml-schema) that can be used
> by editors like [VSCode](/doc/install/plugins#vscode) or
> [PyCharm](/doc/install/plugins#pycharm) to enable automatic syntax
> checks and auto-completion.

Reviewer: ```suggestion
[schema](https://github.com/iterative/dvcyaml-schema) that can be used by
editors like [VSCode](/doc/install/plugins#vscode) or
[PyCharm](/doc/install/plugins#pycharm) to enable automatic syntax checks and
auto-completion.
```
Fixing format (cp from https://github.com/iterative/dvc.org/pull/1686/files)

> in CI.

Reviewer 2: @skshetry would be great to put some links to VS Code and Pycharm is possible on how to install this kind of schema as an example.

Reviewer: It's explained in the README of https://github.com/iterative/pipeline-schema

Reviewer 2: it's better to include a sentence here - easier to search, easier to read, good for SEO, etc.

Reviewer: Maybe in https://dvc.org/doc/install/plugins and link there?

Reviewer 2: it can be a separate entry in the plugins indeed.

Author: @shcheklein @jorgeorpinel, I just wanted to mention it briefly to see if there were any interest and get suggestions from the community before I submit to the catalog of https://github.com/SchemaStore/schemastore.
The installation guide is enough for now, I think. 

> it can be a separate entry in the plugins indeed.

But, this is a good point. We'll anyway need to provide a link to the yaml extension, so I could briefly mention it there (but, probably without mentioning "how-to-install").

Reviewer 2: >  I just wanted to mention it briefly to see if there were any interest and get suggestions from the community before I submit to the catalog of SchemaStore/schemastore

I would say we should submit this - no need to wait. Relying on this small note as a hypothesis test would be not enough - it's hidden too much.

Author: @shcheklein, sure, I'll try to submit this week. Till then, I'll kept this as a draft.

Author: Added to the schemastore. You should be able to just open dvc.yaml and get the schema on Pycharm and VSCode (with YAML extension). For some reason, I'm getting `Connection timed out` on Pycharm, would be great if you guys could confirm that it works without any effort. @shcheklein @jorgeorpinel

Reviewer: Hey @skshetry I see your contribution there (https://github.com/SchemaStore/schemastore/pull/1187/files). Nice, thanks! But I'm not sure how to add or use that on the IDE. On my VSCode, dvc.yaml files just open a YAML files. Same for PyCharm I assume (I don't even have it rn but I've used it before and can download it). Of course I could try and figure it out 😋  but my point is maybe you can write a draft guide in https://dvc.org/doc/install/plugins as mentioned above? I'd be happy to take this over from that point.

Author: For VSCode, you just need to install [YAML](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml) extension and just reload/restart the editor. Then, open `dvc.yaml` file. Based on your network connection, it might take some time but try typing "st" and `Ctrl + Space` to force suggestions.

For pycharm, I'm having some trouble (sometime it works automatically, sometime it just throws me error). For pycharm, 
open a dvc.yaml file. Then, you should get either "No schema available" or "Schema: dvcyaml` written on the bottom status bar. If on "No schema available", click on it and select "dvcyaml", otherwise you are good to go and start typing.

> If you do not see "dvcyaml" in the list, you might need to opt for auto-download schema.

Reviewer: > try typing "st" and Ctrl + Space

Nice, this works on VSCode. Didn't even need the RedHAT YAML extension, it's supported out-of-box it seems.

Reviewer: Ah never mind you were right, I needed the YAML extension. The previous autocomplete was basic IntelliSense based on existing words in the doc.

> in CI.

Reviewer: Questions that may need explaining/ clarifying here:
- What's are the main goals for publishing the JSON schema? Command line validation/linting tools tools, CI checks, IDE auto-completion?
- Why is it a JSON schema if the file format is YAML?

Also, should we rename the repo `dvcyaml-schema` instead of `pipeline-schema`? We want to avoid synonyms so we never use the term "pipelines file", it's always dvc.yaml.

Reviewer: See also https://github.com/iterative/pipeline-schema/issues/4 🙂

Author: > What are the main goals for publishing the JSON schema?

It's the autocompletion and validation/linting inside text editors for now. 

> Why is it a JSON schema if the file format is YAML?

Most of the editors require extensions that use Redhat's YAML language server, which uses JSON schema to validate, because:
1. JSON is a subset of YAML.
2. It's just a schema, so if any format is largely compatible with JSON, it can be described through that schema (though, it might be a bit restrictive).
3. Describing YAML with a different set of the schema (let's say, YAML Schema) is a ton of work, given the fact that JSON Schema is still not standardized.

Reviewer 2: I agree with @jorgeorpinel , JSON will be confusing. Docs are not about being mathematically correct (especially docs outside command reference). Their goal is to provide useful information to users in the most accessible way.

We need to simplify the paragraph. It can be something like:

💡 There is a `dvc.yaml` [schema](link) available that could be installed into editors like [VSCode](link) or [PyCharm](link) to enable automatic syntax checks and autocompletion.

> We have published the
> [JSON Schema](https://github.com/iterative/pipeline-schema) for `dvc.yaml` files
> that helps code editors and IDEs to provide auto-completion and
> validation/linting, and to verify the proper structure of `dvc.yaml` locally
> and/or in CI.

Reviewer: Just submitting Ivan's suggestion from https://github.com/iterative/dvc.org/pull/1650#discussion_r464799496 formally:
```suggestion
💡 We maintain a `dvc.yaml`
[schema](https://github.com/iterative/pipeline-schema) that can be used
by editors like [VSCode](/doc/install/plugins#vscode) or
[PyCharm](/doc/install/plugins#pycharm) to enable automatic syntax
checks and auto-completion.
```
May need formatting. And the plugins.md sections don't exist yet.

Reviewer: I'll commit this for now.
